### PR TITLE
Rework filesystem handling

### PR DESCRIPTION
**Pull request by matz-e**
Does a few things:

* Clears filesystem interfaces before daemonizing
* Splits the `StorageElement` class by factoring out a `FileSystem` class — now code is split along functionality, with each class having one use-case
* Renames `default` to `alternative`, since the default filesystem interactions are output SE interactions, not input SE ones.

I don't know anymore what I was thinking implementing the FS code in the first place? I hope this iteration makes it easier to read and a little less magical?

### CODE REVIEW

File: lobster/se.py [hunk 1]
```diff
@@ -518,18 +523,28 @@ def _initialize(self, methods):
             else:
                 logger.debug("implementation of master access missing for URL {0}".format(url))
 
+    def deactivate(self):
+        """Clear all storage-element objects.
+
+        Needed when Lobster daemonizes, as filesystem objects may have
```
Comment (annawoodard) [inline lobster/se.py]: :+1: For clear explanation to our future selves.

File: lobster/se.py [hunk 2]
```diff
@@ -19,54 +19,30 @@
...
-    """Weird class to handle all needs of storage implementations.
+class FileSystem(object):
 
-    This class can be used for file system operations after at least one of
-    its subclasses has been instantiated.
+    """Singleton class as an interface for filesystem interactions.
 
-    "One size fits nobody." (T. Pratchett)
+    Needs to be configured before first use, with two lists of
+    `StorageElement` implementations.
```
Comment (annawoodard) [inline lobster/se.py]: Would be nice to clarify here the purpose of having two lists.

File: lobster/se.py [hunk 3]
```diff
@@ -58,18 +58,18 @@ def switch(*args, **kwargs):
         return switch
 
     @classmethod
-    def configure(cls, defaults, systems):
+    def configure(cls, defaults, alternatives):
         cls._defaults = defaults
-        cls._systems = systems
+        cls._fallback = alternatives
```
Comment (annawoodard) [inline lobster/se.py]: Can we consider choosing either `fallback` or `alternative` for the parameter and attribute (how we used to have `systems` and `_systems`? The fewer vocab mappings we have, the easier things are to keep straight.

File: lobster/se.py [hunk 4]
```diff
@@ -58,18 +58,18 @@ def switch(*args, **kwargs):
...
 
     @contextmanager
-    def default(self):
-        tmp = FileSystem._systems
-        FileSystem._systems = FileSystem._defaults
+    def alternative(self):
```
Comment (annawoodard) [inline lobster/se.py]: Same here

File: lobster/commands/process.py [hunk 5]
```diff
@@ -148,6 +149,19 @@ def run(self, args):
...
+        openconns = process.connections()
+
+        if len(openconns) > 0 or len(openfiles) > 0:
+            logger.error("cannot daemonize due to open files or connections")
+            for f in openfiles:
+                logger.error("open file: {}".format(f.path))
+            for c in openconns:
+                logger.error("open connection: {}".format(c))
+            raise RuntimeError("open files or connections")
+
```
Comment (annawoodard) [inline lobster/commands/process.py]: :+1::+1::+1::+1::+1::+1::+1::+1::+1::+1::+1::+1::+1:

Comment (klannon) [inline lobster/commands/process.py]: I think @btovar is suggesting that we sleep before doing this check to keep from flagging as open connections that take a little time to close.  Did I understand that correctly from Slack?

https://ndlobster.slack.com/archives/cctools-debug/p1470247696000020

Comment (matz-e) [inline lobster/commands/process.py]: Yes. I'll see about that.

File: lobster/se.py [hunk 6]
```diff
@@ -503,35 +512,41 @@ def _initialize(self, methods):
...
-                Local(path)
+                yield Local(path)
             elif protocol == 'hdfs':
                 try:
-                    Hadoop(path)
+                    yield Hadoop(path)
                 except NameError:
                     raise NotImplementedError("hadoop support is missing on this system")
             elif protocol == 'srm':
                 SRM(url)
```
Comment (klannon) [inline lobster/se.py]: Why no `yield` here?

Comment (matz-e) [inline lobster/se.py]: Oversight m(

**Comment by matz-e** [comment]
Will post this again a bit reworked…

**Pull request closed without merging**